iroha_p2p/src/network.rs

```diff
@@ -0,0 +1,123 @@
+use iroha_actor::broker::Broker;
+use iroha_crypto::{PublicKey};
+use iroha_logger::warn;
+
+use crate::peer::{Peer, State};
+use iroha_data_model::peer::prelude::PeerId;
+use ursa::encryption::symm::Encryptor;
+
+pub struct Network<E: Encryptor + Send + 'static> {
+    peers: HashMap<PeerId, Addr<Peer<E>>>,
+    listener: Option<TcpListener>,
+    addr: Option<Addr<Self>>,
+    broker: Broker,
+    //handler: AlwaysAddr<A>
+}
+
+impl<E: Encryptor + Send + 'static> Network<E> {
+    pub fn new(broker: Broker, /*handler: AlwaysAddr<A>*/) -> Self {
+        Self { peers: HashMap::new(), listener: None, addr: None, broker, /*handler*/ }
+    }
+
+    fn listener_stream(&mut self) -> impl Stream<Item = NewPeerResult> + Send + 'static {
+        let listener = self.listener.take().unwrap();
+        stream! {
+            loop {
+                match listener.accept().await {
+                    Ok((stream, addr)) => {
+                        let id = PeerId { address: addr.to_string(), public_key: PublicKey::default() };
+                        let new_peer: NewPeerResult = NewPeerResult::Ok { stream, id };
+                        yield new_peer;
+                    },
+                    Err(error) => {
+                        yield NewPeerResult::Err { error };
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    async fn connect_peer(&mut self, id: PeerId) {
+        let my_addr = self.addr.clone().expect("Error connecting peer from unstarted Network!");
+        let addr = Peer::new(id.clone(), None, State::Connecting, my_addr).start().await;
+        self.peers.insert(id, addr);
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Actor for Network<E> {
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        self.addr = Some(ctx.addr());
+        // from peer
+        self.broker.subscribe::<ReceivedPeerMsg, _>(ctx);
+        // from other iroha subsystems
+        self.broker.subscribe::<SendPeerMsg, _>(ctx);
+        // from listener
+        ctx.notify_with(self.listener_stream());
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Handler<SendPeerMsg> for Network<E> {
+    type Result = ();
+
+    async fn handle(&mut self, msg: SendPeerMsg) {
+        let addr: &Addr<Peer<E>> = &self.peers[&msg.id];
+        addr.do_send(msg).await;
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Handler<ReceivedPeerMsg> for Network<E> {
+    type Result = ();
+
+    async fn handle(&mut self, msg: ReceivedPeerMsg) {
+        // TODO send message to Torii
+        //self.handler.do_send(msg.message);
+    }
+}
+
+#[async_trait::async_trait]
+impl<E: Encryptor + Send + 'static> Handler<NewPeerResult> for Network<E> {
+    type Result = ();
+
+    async fn handle(&mut self,  peer: NewPeerResult) {
+        let addr = self.addr.clone().unwrap();
```

Comment:
Handler has context as second argument. You dont need to carry around your own address. You can get it from context.

Reply:
Converted Handler to ContextHandler here.